Src/Support/Google.Apis.Auth/OAuth2/Flows/PkceGoogleAuthorizationCodeFlow.cs

```diff
@@ -0,0 +1,99 @@
+    {
+        private const int CodeVerifierMaxLengthChars = 128;
+        // We generate the random bytes and then encode them to Base64 whose characters are 6 bits long.
+        private const int CodeVerifierRandomNumberLengthBytes = CodeVerifierMaxLengthChars * 6 / 8;
+        private const string ChallengeMethodSha256 = "S256";
+        private const string ChallengeMethodPlain = "plain";
+
+        /// <summary>
+        /// Creates a new instance from the given initializer.
+        /// </summary>
+        public PkceGoogleAuthorizationCodeFlow(Initializer initializer) : base(initializer)
+        {
+        }
+
+        /// <inheritdoc/>
+        public AuthorizationCodeRequestUrl CreateAuthorizationCodeRequest(string redirectUri, out string codeVerifier)
+        {
+            var request = CreateAuthorizationCodeRequest(redirectUri) as GoogleAuthorizationCodeRequestUrl;
+            codeVerifier = GenerateCodeVerifier();
+            AddCodeChallenge(request, codeVerifier);
+            return request;
+        }
+
+        /// <inheritdoc/>
+        public async Task<TokenResponse> ExchangeCodeForTokenAsync(string userId, string code, string codeVerifier, string redirectUri, CancellationToken taskCancellationToken)
+        {
+            var request = CreateAuthorizationCodeTokenRequest(userId, code, redirectUri);
+            request.CodeVerifier = codeVerifier;
+
+            return await ExchangeCodeForTokenAsync(userId, request, taskCancellationToken).ConfigureAwait(false);
+        }
+
+        private static string GenerateCodeVerifier()
+        {
+            byte[] data = new byte[CodeVerifierRandomNumberLengthBytes];
+            using RandomNumberGenerator randomNumberGenerator = RandomNumberGenerator.Create();
+            randomNumberGenerator.GetBytes(data);
+
+            // The code verifier alphabet does not allow +, / or =, so we Base64URL encode.
+            // https://developers.google.com/identity/protocols/oauth2/native-app#step1-code-verifier
+            return TokenEncodingHelpers.UrlSafeBase64Encode(data);
+        }
+
+        private static void AddCodeChallenge(GoogleAuthorizationCodeRequestUrl request, string codeVerifier)
+        {
+            // https://developers.google.com/identity/protocols/oauth2/native-app#step1-code-verifier
+            string effectiveChallenge;
+            string effectiveChallengeMethod;
+            try
+            {
+                byte[] codeVerifierAsciiBytes = Encoding.ASCII.GetBytes(codeVerifier);
+                using var sha256 = SHA256.Create();
+                byte[] hashedCodeVerifier = sha256.ComputeHash(codeVerifierAsciiBytes);
+
+                effectiveChallenge = TokenEncodingHelpers.UrlSafeBase64Encode(hashedCodeVerifier);
+                effectiveChallengeMethod = ChallengeMethodSha256;
+            }
+            catch (TargetInvocationException)
```

Comment:
Possibly add a comment here as you did in the GitHub conversation to explain why this might happen?

Reply:
Yes, done.